File: sdks/python/src/ankka/effects/event_sourced.py
"""Effects for event sourced entities: persist events, then decide what to reply.

State can only change by persisting an event — there is deliberately no ``update_state`` here,
mirroring the Scala SDK. A ``ReadOnlyEffect`` provably persists nothing; ``@query`` handlers
must return one, and the sidecar refuses events from a read-only handler regardless.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from datetime import timedelta
from typing import Any, Generic, TypeVar

from ankka.context import Metadata
from ankka.effects.common import DeleteNow, Error, ErrorCode, ExpireAfter, Fail, NoReply, Outcome, Reply, Retention

S = TypeVar("S")
E = TypeVar("E")
R = TypeVar("R")


@dataclass(frozen=True)
class EventSourcedEffect(Generic[S, E, R]):
    events: tuple[E, ...]
    retention: Retention | None
    outcome: Outcome


@dataclass(frozen=True)
class ReadOnlyEffect(EventSourcedEffect[S, E, R]):
    """An effect that persists nothing. Its ``events`` are always empty."""

    events: tuple[E, ...] = ()
    retention: Retention | None = None
    outcome: Outcome = NoReply()


@dataclass(frozen=True)
class PersistBuilder(Generic[S, E]):
    """Events and retention accumulated before the reply is chosen."""

    events: tuple[E, ...]
    retention: Retention | None = None

    def delete_entity(self) -> PersistBuilder[S, E]:
        return PersistBuilder(self.events, DeleteNow())

    def expire_after(self, duration: timedelta) -> PersistBuilder[S, E]:
        return PersistBuilder(self.events, ExpireAfter(duration))

    def then_reply(self, compute: Callable[[S], R], metadata: Metadata | None = None) -> EventSourcedEffect[S, E, R]:
        """Reply with a value derived from the state *after* the events are applied."""
        return EventSourcedEffect(self.events, self.retention, Reply(compute, metadata or Metadata()))

    def then_reply_state(self) -> EventSourcedEffect[S, E, S]:
        return EventSourcedEffect(self.events, self.retention, Reply(lambda s: s))

    def then_no_reply(self) -> EventSourcedEffect[S, E, Any]:
        return EventSourcedEffect(self.events, self.retention, NoReply())


class EventSourcedEffects(Generic[S, E]):
    """The ``effects`` surface inside an event sourced entity's handlers. Stateless."""

    def persist(self, event: E, *more: E) -> PersistBuilder[S, E]:
        return PersistBuilder((event, *more))

    def persist_all(self, events: list[E] | tuple[E, ...]) -> PersistBuilder[S, E]:
        return PersistBuilder(tuple(events))

    def reply(self, value: R, metadata: Metadata | None = None) -> ReadOnlyEffect[S, E, R]:
        return ReadOnlyEffect(outcome=Reply(lambda _: value, metadata or Metadata()))

    def error(self, message: str, code: ErrorCode = ErrorCode.BAD_REQUEST) -> ReadOnlyEffect[S, E, Any]:
        return ReadOnlyEffect(outcome=Fail(Error(message, code)))

    def no_reply(self) -> ReadOnlyEffect[S, E, Any]:
        return ReadOnlyEffect(outcome=NoReply())

    def delete_entity(self) -> PersistBuilder[S, E]:
        """Delete the entity without recording a final event."""
        return PersistBuilder((), DeleteNow())

File: sdks/typescript/src/effects/eventSourced.ts
// The effects an event sourced entity's handlers return: persist these events and then reply from the
// state after them, or answer without persisting. Inert values. `ReadOnlyEffect` is a distinct type
// with `kind: "read-only"`, which is what lets `query` refuse a persisting handler at compile time.

import type { Duration } from "../time.ts"
import { done, type Done } from "../schema.ts"
import { ErrorCode, Outcome, Retention, type EffectLike, type Metadata, type ReadOnlyLike } from "./common.ts"

export interface PersistEffect<S, E, R> extends EffectLike<R> {
  readonly kind: "persist"
  readonly events: readonly E[]
  readonly retention: Retention | null
  readonly outcome: Outcome<S, R>
}

export interface ReadOnlyEffect<S, E, R> extends ReadOnlyLike<R> {
  readonly kind: "read-only"
  readonly retention: Retention | null
  readonly outcome: Outcome<S, R>
  /** Phantom: the entity's event type, so a read-only effect is tied to its entity. */
  readonly _events?: E
}

export type EventSourcedEffect<S, E, R> = PersistEffect<S, E, R> | ReadOnlyEffect<S, E, R>

/** `effects.persist(e)` returns one of these; finish it with `thenReply`, `thenReplyState` or `thenNoReply`. */
export class PersistBuilder<S, E> {
  readonly #events: readonly E[]
  readonly #retention: Retention | null

  constructor(events: readonly E[], retention: Retention | null = null) {
    this.#events = events
    this.#retention = retention
  }

  /** Delete the entity once these events are persisted. */
  deleteEntity(): PersistBuilder<S, E> {
    return new PersistBuilder(this.#events, Retention.deleteNow)
  }

  /** Expire the entity a while after these events are persisted. */
  expireAfter(after: Duration): PersistBuilder<S, E> {
    return new PersistBuilder(this.#events, Retention.expireAfter(after))
  }

  /** Reply with a value computed from the state *after* the events. */
  thenReply<R>(compute: (state: S) => R, metadata?: Metadata): PersistEffect<S, E, R> {
    return Object.freeze({ kind: "persist", events: this.#events, retention: this.#retention, outcome: Outcome.reply(compute, metadata) })
  }

  /** Reply with the state after the events. */
  thenReplyState(): PersistEffect<S, E, S> {
    return this.thenReply((state) => state)
  }

  /** Persist and answer nothing. */
  thenNoReply(): PersistEffect<S, E, never> {
    return Object.freeze({ kind: "persist", events: this.#events, retention: this.#retention, outcome: Outcome.noReply() })
  }
}

/** The factory a handler reaches as `this.effects`. */
export class EventSourcedEffects<S, E> {
  /** Persist one or more events. */
  persist(event: E, ...more: E[]): PersistBuilder<S, E> {
    return new PersistBuilder<S, E>(Object.freeze([event, ...more]))
  }

  /** Persist a list of events; an empty list is allowed and persists nothing. */
  persistAll(events: readonly E[]): PersistBuilder<S, E> {
    return new PersistBuilder<S, E>(Object.freeze([...events]))
  }

  /** Answer without persisting. */
  reply<R>(value: R, metadata?: Metadata): ReadOnlyEffect<S, E, R> {
    return Object.freeze({ kind: "read-only", retention: null, outcome: Outcome.reply(() => value, metadata) })
  }

  /** Refuse the command: nothing is persisted and the caller sees the code. */
  error(message: string, code: ErrorCode = ErrorCode.BadRequest): ReadOnlyEffect<S, E, never> {
    return Object.freeze({ kind: "read-only", retention: null, outcome: Outcome.fail(message, code) })
  }

  /** Answer nothing and persist nothing. */
  noReply(): ReadOnlyEffect<S, E, never> {
    return Object.freeze({ kind: "read-only", retention: null, outcome: Outcome.noReply() })
  }

  /** Delete the entity, persisting nothing more; finish with `thenReply(() => done)` or `thenNoReply()`. */
  deleteEntity(): PersistBuilder<S, E> {
    return new PersistBuilder<S, E>(Object.freeze([]), Retention.deleteNow)
  }

  /** Expire the entity after a while, persisting nothing more. */
  expireAfter(after: Duration): PersistBuilder<S, E> {
    return new PersistBuilder<S, E>(Object.freeze([]), Retention.expireAfter(after))
  }
}

export { done, type Done }
